src/dotty/tools/dotc/core/SymbolLoaders.scala

```diff
@@ -86,18 +86,27 @@ class SymbolLoaders {
       (module, modcls) => new PackageLoader(module, pkg)).entered
   }
 
+  /** Whether the class to be loaded is `java.lang.Object`
+   *
+   *  `java.lang.Object` should never be loaded from classfiles
+   */
+  def isObjectClass(owner: Symbol, name: PreName)(implicit ctx: Context): Boolean =
+    owner.fullName.toString == "java.lang" && name.show == "Object"
+
   /** Enter class and module with given `name` into scope of `owner`
    *  and give them `completer` as type.
    */
   def enterClassAndModule(
       owner: Symbol, name: PreName, completer: SymbolLoader,
       flags: FlagSet = EmptyFlags, scope: Scope = EmptyScope)(implicit ctx: Context): Unit = {
     val clazz = enterClass(owner, name, completer, flags, scope)
-    val module = enterModule(
-      owner, name, completer,
-      modFlags = flags.toTermFlags & RetainedModuleValFlags,
-      clsFlags = flags.toTypeFlags & RetainedModuleClassFlags,
-      scope = scope)
+
```

Comment:
Note: The `NoType` technique is not an accident. It is used for all companion objects that are entered (because we see their file name) but turn out to not exist after class loading. Here we have to do it before class loading because once we try to load Object's classfile it is already too late.